File: projects/organic04/java/FadingColor.java
import java.awt.Color;
public class FadingColor {
	private Color start, current, end;
	private int steps;
	private Color[] sequence;
	private double	deltaR,
						deltaG,
						deltaB,
						doubleR,
						doubleG,
						doubleB;
	private boolean finished = false;
	private int index = 0;
	
	
	public FadingColor(Color start, Color end, int steps) {
		this.start = start;
		this.end = end;
		this.steps = steps;
		this.sequence = new Color[steps];
		reset();
	}
	
	public Color at(int i) {
		try {
			return sequence[i];
		} catch(NullPointerException e) {
			return end;
		}
	}
	
	public Color next() {
		try {
			return sequence[index];
		} catch(NullPointerException e) {
			return end;
		}
	}
	public boolean hasNext() {
		return (index < steps);
	}
	
	private void reset() {
		sequence = new Color[steps];
		setDeltas();
		for(int i = 0; i < steps; i++) {
			doubleR += deltaR;
			doubleG += deltaG;
			doubleB += deltaB;
			int r = (int)(Math.round(doubleR));
			int g = (int)(Math.round(doubleG));
			int b = (int)(Math.round(doubleB));
			sequence[i] = new Color(r,g,b);
		}
		finished = false;
		index = 0;
		
		for(int i = 0; i < steps; i++) {
			int r = sequence[i].getRed();
			int g = sequence[i].getGreen();
			int b = sequence[i].getBlue();
		}
	}
	
	private void setDeltas() {
		deltaR = (double)(start.getRed() - end.getRed()) / steps * -1;
		deltaG = (double)(start.getGreen() - end.getGreen()) / steps * -1;
		deltaB = (double)(start.getBlue() - end.getBlue()) / steps * -1;
		
		doubleR = (double)start.getRed();
		doubleG = (double)start.getGreen();
		doubleB = (double)start.getBlue();
	}
}

File: archives/frontpages/07/SCRIPTS/FadingColor.js
function FadingColor (startColor, endColor, numberOfSteps) {
	if (startColor.isColor && endColor.isColor) {
		this.startColor = startColor;
		this.currentColor = this.startColor;
		this.endColor = endColor;
		this.numberOfSteps = numberOfSteps;
		
		this.ready = false;
		this.finished = false;
		this.deltaR = this.deltaG = this.deltaB = this.doubleR = this.doubleG = this.doubleB = 0.0;
		
		this.makeReady = makeReady;
		this.makeFinished = makeFinished;
		this.setDeltas = setDeltas;
		this.swap = swap;
		this.isFinished = isFinished;
		this.getNextColor = getNextColor;
		
		this.makeReady ();
	}
	
	function isFinished () { return this.finished; }
	
	function makeReady () {
		this.setDeltas ();
		this.ready = true;
		this.finished = false;
	}
	function makeFinished () {
		this.ready = false;
		this.finished = true;
	}
	function swap () {
		var t = this.startColor;
		this.startColor = this.endColor;
		this.endColor = t;
	}
	
	function setDeltas () {
		this.deltaR = (this.startColor.r - this.endColor.r) / this.numberOfSteps * -1;
		this.deltaG = (this.startColor.g - this.endColor.g) / this.numberOfSteps * -1;
		this.deltaB = (this.startColor.b - this.endColor.b) / this.numberOfSteps * -1;
		this.doubleR = this.startColor.r;
		this.doubleG = this.startColor.g;
		this.doubleB = this.startColor.b;
	}
	
	function getNextColor () {
		this.doubleR += this.deltaR;
		this.doubleG += this.deltaG;
		this.doubleB += this.deltaB;
		var c = new Color (Math.round (this.doubleR), Math.round (this.doubleG), Math.round (this.doubleB));
		if (c.equals (this.endColor)) {
			this.makeFinished ();
		}
		return c;
	}
}